src/app/Plans/components/Wizard/MappingForm.tsx

```diff
@@ -0,0 +1,143 @@
+import {
+  Checkbox,
+  Grid,
+  GridItem,
+  Radio,
+  TextInput,
+  Title,
+  Form,
+  FormGroup,
+  Flex,
+  FlexItem,
+} from '@patternfly/react-core';
+import spacing from '@patternfly/react-styles/css/utilities/Spacing/spacing';
+
+import SimpleSelect, { OptionWithValue } from '@app/common/components/SimpleSelect';
+import { IStorageMapping, MappingType, INetworkMapping } from '@app/Mappings/types';
+import { MappingBuilder, IMappingBuilderItem } from '@app/Mappings/components/MappingBuilder';
+import { MappingSource, MappingTarget } from '@app/Mappings/types';
+
+interface IMappingFormProps {
+  mappingType: MappingType;
+  mappingList: IStorageMapping[] | INetworkMapping[];
+  availableSources?: MappingSource[];
+  availableTargets?: MappingTarget[];
+}
+
+const MappingForm: React.FunctionComponent<IMappingFormProps> = ({
+  mappingType,
+  mappingList,
+  availableSources = [],
+  availableTargets = [],
+}: IMappingFormProps) => {
+  const mappingOptions = Object.values(mappingList).map((mapping) => ({
+    toString: () => mapping.name,
+    value: mapping,
+  })) as OptionWithValue<IStorageMapping | INetworkMapping>[];
+
+  const [isCreateMappingChecked, toggleCreateMapping] = React.useReducer(
+    (isCreateMappingChecked) => !isCreateMappingChecked,
+    false
+  );
+
+  const [isSaveNewMapping, toggleSaveNewMapping] = React.useReducer(
+    (isSaveNewMapping) => !isSaveNewMapping,
+    false
+  );
+
+  const [mapping, setMapping] = React.useState<IStorageMapping | INetworkMapping | null>(null);
+  const [mappingName, setMappingName] = React.useState<string>('');
+
+  // TODO add support for prefilling builderItems for editing an API mapping
+  const [builderItems, setBuilderItems] = React.useState<IMappingBuilderItem[]>([
+    { source: null, target: null },
+  ]);
+
+  return (
+    <Form>
+      <Title headingLevel="h3" size="md">
+        Select an existing {mappingType.toLowerCase} mapping between your source and target
+        providers, or create a new one.
+      </Title>
+      <Flex direction={{ default: 'column' }} className={spacing.mbMd}>
+        <Radio
+          id="existing-mapping"
+          label={`Use an existing ${mappingType.toLowerCase()} mapping`}
+          name="radio-existing"
+          onChange={toggleCreateMapping}
+          isChecked={!isCreateMappingChecked}
+        />
+        {!isCreateMappingChecked && (
+          <FlexItem>
+            <FormGroup
+              isRequired
+              fieldId="sourceProvider"
+              helperTextInvalid="TODO"
+              // TODO add state/validation/errors to this and other FormGroups
+              validated="default"
+            >
+              <SimpleSelect
+                id="mappingSelect"
+                options={mappingOptions}
+                value={[mappingOptions.find((option) => option.value === mapping)]}
+                onChange={(selection) =>
+                  setMapping((selection as OptionWithValue<IStorageMapping>).value)
```

Comment:
This should probably be `as OptionWithValue<Mapping>` or `as OptionWithValue<IStorageMapping | INetworkMapping>`, since it's storage-specific now. Eventually I hope to add generics to `SimpleSelect` though so we don't have to assert this type at all.